src/ProxyProtocol.cc

```diff
@@ -0,0 +1,325 @@
+
+        if (!message->srcIpAddr.GetHostByName(ipa.c_str()))
+            throw TexcHere("PROXY/1.0 error: invalid src-IP address");
+
+        if (!message->dstIpAddr.GetHostByName(ipb.c_str()))
+            throw TexcHere("PROXY/1.0 error: invalid dst-IP address");
+
+        if (porta > 0 && porta <= 0xFFFF) // max uint16_t
+            message->srcIpAddr.port(static_cast<uint16_t>(porta));
+        else
+            throw TexcHere("PROXY/1.0 error: invalid src port");
+
+        if (portb > 0 && portb <= 0xFFFF) // max uint16_t
+            message->dstIpAddr.port(static_cast<uint16_t>(portb));
+        else
+            throw TexcHere("PROXY/1.0 error: invalid dst port");
+
+    } else if (contentTok.skip(unknown)) {
+        // found valid but unusable header
+        // discard the rest of the line
+        message->protoSupported = false;
+    } else
+        throw TexcHere("PROXY/1.0 error: invalid INET protocol or family");
+
+    buf.consume(tok.parsedSize());
+    return message;
+}
+
+static ProxyProtocol::Message::Pointer
+ParseV2(SBuf &buf)
+{
+    static const SBuf::size_type magicLength = Proxy2p0magic.length();
+    static const SBuf::size_type minMessageLength = magicLength + 4;
+
+    ProxyProtocol::Message::Pointer message;
+
+    ::Parser::BinaryTokenizer tok(buf, true);
+    tok.skip(magicLength, "magic");
+
+    const auto versionAndCommand = tok.uint8("version and command");
+
+    const auto version = (versionAndCommand & 0xF0) >> 4;
+    if (version != 2) // version == 2 is mandatory
+        throw TexcHere(ToSBuf("PROXY/2.0 error: invalid version ", version));
+
+    const auto command = (versionAndCommand & 0x0F);
+    if ((command & 0xFE) != 0x00) // values other than 0x0-0x1 are invalid
+        throw TexcHere(ToSBuf("PROXY/2.0 error: invalid command ", command));
+
+    debugs(88, 3, "parsed pp2_tlv command " << command);
+
+    const auto familyAndProto = tok.uint8("family and proto");
+
+    const auto family = (familyAndProto & 0xF0) >> 4;
+    if (family > 0x3) // values other than 0x0-0x3 are invalid
+        throw TexcHere(ToSBuf("PROXY/2.0 error: invalid address family ", family));
+
+    const auto proto = (familyAndProto & 0x0F);
+    if (proto > 0x2) // values other than 0x0-0x2 are invalid
+        throw TexcHere(ToSBuf("PROXY/2.0 error: invalid transport protocol ", proto));
+
+    // the header length field contains the number following bytes beyond this field
+    auto headerLen = tok.uint16("header length");
+
+    if (headerLen > std::numeric_limits<uint16_t>::max() - minMessageLength)
+        throw TexcHere(ToSBuf("PROXY/2.0 error: an invalid header length: expecting an integer less than ",
+                    std::numeric_limits<uint16_t>::max() - minMessageLength, " but got ", headerLen));
+
+    // the total message length, including the magic bytes
+    headerLen += minMessageLength;
+
+    message = new ProxyProtocol::Message("2.0", command);
+
+    message->protoSupported = (proto != 0);
```

Comment:
The PROXY specs say that for UNSPEC protocol we should discard the address block. Should we really do that? Probably we should handle this case as "local connections" below? In this case, the TLV part would be also discarded.

Reply:
> The PROXY specs say that for UNSPEC protocol we should discard the address block. Should we really do that?

Yes, of course. We cannot parse addresses if we do not know what protocol they belong to. Same for AF_UNSPEC.

Please note that yours (and official) code relies on family to determine the address type. AFAICT, it should instead rely on familyAndProto _combination_ as documented in the PROXY protocol specs. After fixing this, we should mention this fix in the PR description.

Reply:
IMO, the TLVs part should be ignored if (and only if) we cannot find the start and size of the TLVs part. If finding TLVs requires knowing address sizes, then we have to ignore TLVs (if any) when we do not know address sizes.

Again, if that logic changes how Squid handles incoming PROXY, we should document the change and its expected high-level effects in the PR description.

Reply:
Done at 59c1e38.